### PR TITLE
compose: Don't send immediately on adding image

### PR DESCRIPTION
**Pull request by chrisbobbe**
Thanks @AkashDhiman for your PR #5125 and your prep work for this in #4590; those were quite helpful and informative. 🙂

Marked as a draft/blocked for a reason I'll explain, but that reason doesn't affect the first three commits—

ComposeMenu [nfc]: Use our own new type definition for attachments
ComposeMenu [nfc]: Fix name of insertAttachment prop to be plural
ComposeMenu [nfc]: Convert to a function component

—and indeed those three would be good to land as prep for removing the blocker.

The blocker is that we have basically the behavior of #5291, except with the insertion of image links instead of Jitsi links. On iOS, you take or select a photo to upload, and sometimes the message text input is unchanged when you expect it to get an inserted `[Uploading …]()` link.

Fixes: #2366
Fixes: #4540
Replaces: #5125

### CODE REVIEW

**Comment by gnprice** [comment]
Chat thread about the resulting UI on Android:
https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/Android.20select.20multiple.20photos/near/1423109

File: src/compose/ComposeMenu.js [hunk 1]
```diff
@@ -84,90 +77,128 @@ export const chooseUploadImageFilename = (uri: string, fileName: string): string
   return nameWithoutPrefix;
 };
 
-class ComposeMenuInner extends PureComponent<Props> {
-  static contextType = TranslationContext;
-  context: GetText;
+export default function ComposeMenuInner(props: Props): Node {
```
Comment (gnprice) [inline src/compose/ComposeMenu.js:80]: nit: There's no longer a distinct "inner" vs "outer" component here, right? So this function should be called just `ComposeMenu`.

(Another angle on the same thing: it's the default export, so should have the same name as the file.)

File: src/compose/ComposeMenu.js [hunk 2]
```diff
@@ -84,90 +77,128 @@ export const chooseUploadImageFilename = (uri: string, fileName: string): string
...
-      return;
-    }
 
-    dispatch(uploadFile(destinationNarrow, uri, chooseUploadImageFilename(uri, fileName)));
-  };
+      if (numMalformed > 0) {
+        let msg = undefined;
+        if (assets.length === 1 && numMalformed === 1) {
+          msg = _('Failed to attach your file.');
+          return;
```
Comment (gnprice) [inline src/compose/ComposeMenu.js:161]: This sets a local and then returns -- so the assignment has no effect, as nothing else can see it.

Perhaps just put the `showErrorAlert` call here? It's pretty concise, not much more complex than the variable assignment.

**Review by gnprice — COMMENTED** [review]
Thanks @chrisbobbe! This code all looks good except the small points below.

That leaves as the interesting questions:
* The expansion of #5291. I think the plan of merging the ComposeMenu refactors from the start of this branch, then going for #5141, makes sense.
* The UI that results on Android. One hopes to continue getting a nice photo-gallery-style UI, rather than a file-oriented UI.

  ISTR we found that newer Android versions, including what one can get in the emulator, do better here and the file-oriented UI might be limited to older versions like in the physical test device you have.

**Comment by chrisbobbe** [comment]
Thanks for the review! Revision pushed.

**Comment by chrisbobbe** [comment]
@gnprice, I've just rebased this atop #5635, so it's ready for another code review. I'm working on spinning up the simulator to make a screen recording now. 🙂

**Comment by chrisbobbe** [comment]
Just pushed a revision fixing a rebase error in the new type `Attachment`'s definition (`r-n-document-picker` changed an optional `string` property to required with `string | null`).

Here's a video of the feature working on my iPhone:

https://user-images.githubusercontent.com/22248748/211920752-a6325fb5-5c08-4c13-b113-8b778ef2ebd0.mp4

File: src/compose/ComposeMenu.js [hunk 3]
```diff
@@ -140,28 +142,54 @@ export default function ComposeMenu(props: Props): Node {
...
-          'uri == null': uri == null,
-          'fileName == null': fileName == null,
+        showErrorAlert(_('Error'), _('Failed to attach your file.'));
+        logging.error('Image picker response gave falsy `assets` or falsy `assets[0]`', {
+          '!assets': !assets,
         });
         return;
       }
 
-      dispatch(uploadFile(destinationNarrow, uri, chooseUploadImageFilename(uri, fileName)));
```
Comment (gnprice) [inline src/compose/ComposeMenu.js:162]: Looks like this was the only call site of that `uploadFile` function — so let's cut that out too.

**Review by gnprice — COMMENTED** [review]
Thanks for the revision!

Code all looks good; small comment below.

On the UI this gets us on Android: I'm still not a fan of that file-picker-style UI when the user asks to upload a photo. The first part of this branch, the "don't send immediately", is a major functionality improvement. But "select multiple images at once" doesn't seem as necessary:
* Lots of messaging apps don't even let you send multiple images in one message.
* If you do want to send multiple images, then given the "don't send immediately" part of this branch, all you have to do is go through the selection multiple times.

So I'm not sure that that part is a good trade for the added inconvenience of the file-picker flow each time you want to send one photo.

Therefore I think what I want to do here is:
* Let's merge the first part of this branch: the "don't send immediately", plus the prep commits for "select multiple images", so in total all but the last commit.
  * With that one code adjustment mentioned below.
* Then for #2366… hmm, looking at it, both of the actual user requests there (the #2366 OP, plus the report at https://chat.zulip.org/#narrow/stream/48-mobile/topic/Bug.20reports/near/1292530 ) are about *sending* multiple images at once, not *selecting* from the picker all at once. Moreover, both of them are really mostly about "don't send immediately, let me add some text", i.e. #4540.
  So I think it would actually be appropriate to have the "Don't immediately send" commit close #2366 as well as #4540. Then we can have a fresh issue thread for anything further we want to do here.
* And if the code you've already written in that last commit gives a UX improvement on iOS, then we can follow up with a version of that commit that's tweaked to be conditional on the platform. No need to spend further effort on the Android side — I think we've already determined there's no *easy* solution there, and the "select multiple images at a time" functionality doesn't seem important enough to justify a bigger effort.

**Comment by chrisbobbe** [comment]
Thanks for the review! Revision pushed.

**Comment by gnprice** [comment]
Thanks! Looks good; merging.

**Pull request merged**